src/backend/gporca/libgpopt/src/xforms/CXformIndexGet2IndexOnlyScan.cpp

```diff
@@ -0,0 +1,168 @@
+//---------------------------------------------------------------------------
+//	Greenplum Database
+//	Copyright (C) 2020 VMware, Inc.
+//
+//	@filename:
+//		CXformIndexGet2IndexOnlyScan.cpp
+//
+//	@doc:
+//		Implementation of transform
+//---------------------------------------------------------------------------
+
+#include <cwchar>
+
+#include "gpos/base.h"
+#include "gpopt/xforms/CXformIndexGet2IndexOnlyScan.h"
+#include "gpopt/xforms/CXformUtils.h"
+
+#include "gpopt/operators/CExpressionHandle.h"
+#include "gpopt/operators/CLogicalIndexGet.h"
+#include "gpopt/operators/CPatternLeaf.h"
+#include "gpopt/operators/CPhysicalIndexOnlyScan.h"
+#include "gpopt/metadata/CIndexDescriptor.h"
+#include "gpopt/metadata/CTableDescriptor.h"
+#include "naucrates/md/CMDIndexGPDB.h"
+
+using namespace gpopt;
+
+
+//---------------------------------------------------------------------------
+//	@function:
+//		CXformIndexGet2IndexOnlyScan::CXformIndexGet2IndexOnlyScan
+//
+//	@doc:
+//		Ctor
+//
+//---------------------------------------------------------------------------
+CXformIndexGet2IndexOnlyScan::CXformIndexGet2IndexOnlyScan
+	(
+	CMemoryPool *mp
+	)
+	:
+	// pattern
+	CXformImplementation
+		(
+		GPOS_NEW(mp) CExpression
+				(
+				mp,
+				GPOS_NEW(mp) CLogicalIndexGet(mp),
+				GPOS_NEW(mp) CExpression(mp, GPOS_NEW(mp) CPatternLeaf(mp))	// index lookup predicate
+				)
+		)
+{}
+
+CXform::EXformPromise CXformIndexGet2IndexOnlyScan::Exfp
+(
+ CExpressionHandle &exprhdl
+)
+const
+{
+	CLogicalIndexGet *popGet = CLogicalIndexGet::PopConvert(exprhdl.Pop());
+
+	CTableDescriptor *ptabdesc = popGet->Ptabdesc();
+	CIndexDescriptor *pindexdesc = popGet->Pindexdesc();
+
+	if ((pindexdesc->IndexType() == IMDIndex::EmdindBtree && ptabdesc->IsAORowOrColTable()) ||
+		(pindexdesc->IndexType() != IMDIndex::EmdindBtree))
```

Comment:
```suggestion
	if (pindexdesc->SupportsIndexOnlyScan())
```

As a design feedback, this check seems out of place. IMO this logic needs to be shifted into the metadata interface, something along the lines like the above pseudo code. I'm not suggesting a behavior change, but rather moving the logic. This will have the added benefit that any change in how we relax or extend our usage will be confined to the metadata, not this transform.

Reply:
I like your proposal.  Slight wrinkle is that existing condition also uses whether it's an AO/AOCO table to make decision, so it doesn't seem as easy to move the entire decision to the index descriptor.  I left the existing condition as is, but added `SupportsIndexOnlyScan()` interface as you suggested.